File: src/parser_utils/invalidation_api.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Make
    (L : Loc_sig.S)
    (Scope_api : Scope_api_sig.S with module L = L)
    (Ssa_api : Ssa_api.S with module L = L)
    (Provider_api : Provider_api.S with module L = L) =
struct
  let is_const_like info values loc =
    try
      let { Scope_api.Def.locs = def_locs; _ } = Scope_api.def_of_use info loc in
      let uses = Scope_api.uses_of_use info loc in
      (* We consider a binding to be const-like if all reads point to the same
         write, modulo initialization. *)
      let writes =
        L.LSet.fold
          (fun use acc ->
            match L.LMap.find_opt use values with
            | None -> (* use is a write *) acc
            | Some write_locs ->
              (* use is a read *)
              (* collect writes pointed to by the read, modulo initialization *)
              List.fold_left
                (fun acc -> function
                  | Ssa_api.Uninitialized -> acc
                  | Ssa_api.Write reason -> L.LSet.add (Reason.poly_loc_of_reason reason) acc)
                acc
                write_locs)
          uses
          L.LSet.empty
      in
      match L.LSet.choose_opt writes with
      | None -> true
      | Some loc -> L.LSet.cardinal writes <= 1 && Nel.mem ~equal:L.equal loc def_locs
    with
    (* TODO: anywhere where we would raise a Missing_def
       error is likely a bug with the Scope_builder that
       should be fixed for the SSA environment to work *)
    | Scope_api.Missing_def _ -> false

  type initialization_valid =
    | Valid
    | NotWritten
    | NullWritten of L.t

  let declaration_validity info values providers loc =
    match Provider_api.providers_of_def providers loc with
    | None -> Valid
    | Some (true, _ :: _) -> Valid
    | Some (_, providers) ->
      (try
         let null_providers =
           (* Since this variable is not fully initialized, if there are any providers then
              they must be null providers *)
           Base.List.map ~f:Reason.poly_loc_of_reason providers |> L.LSet.of_list
         in
         let uses = Scope_api.uses_of_use info loc in
         let reads_exist =
           L.LSet.exists (fun use -> L.LMap.find_opt use values |> Base.Option.is_some) uses
         in
         match (reads_exist, not (L.LSet.is_empty null_providers)) with
         | (false, _) -> Valid
         | (true, false) -> NotWritten
         | (true, true) -> NullWritten (L.LSet.min_elt null_providers)
       with
      (* TODO: anywhere where we would raise a Missing_def
         error is likely a bug with the Scope_builder that
         should be fixed for the SSA environment to work *)
      | Scope_api.Missing_def _ -> Valid)

  let is_not_captured_by_closure =
    let rec lookup in_current_var_scope info scope def =
      if SMap.exists (fun _ def' -> def = def') scope.Scope_api.Scope.defs then
        in_current_var_scope
      else
        match scope.Scope_api.Scope.parent with
        | None -> true
        | Some scope_id' ->
          let scope' = Scope_api.scope info scope_id' in
          lookup (in_current_var_scope && scope.Scope_api.Scope.lexical) info scope' def
    in
    fun info use ->
      let scopes = info.Scope_api.scopes in
      IMap.fold
        (fun _scope_id scope acc ->
          match L.LMap.find_opt use scope.Scope_api.Scope.locals with
          | Some def -> acc && lookup true info scope def
          | None -> acc)
        scopes
        true

  let written_by_closure info values loc =
    let uses = Scope_api.uses_of_use info loc in
    L.LSet.fold
      (fun use acc ->
        match L.LMap.find_opt use values with
        | None ->
          (* use is a write *)
          if is_not_captured_by_closure info use then
            acc
          else
            L.LSet.add use acc
        | Some _write_locs ->
          (* use is a read *)
          (* collect writes pointed to by the read, modulo initialization *)
          acc)
      uses
      L.LSet.empty

  (* Some variables are unhavocable by making a function call but still can be havoced in other
     situations. `via_call` indicates whether the havocing operation is a call or something
     else (resetting an activation scope, yielding, entering a new scope *)

  let mk_caches () = (ref L.LMap.empty, ref L.LMap.empty)

  let should_invalidate ~all (const_like_cache, written_by_closure_cache) info values loc =
    let const_like =
      match L.LMap.find_opt loc !const_like_cache with
      | Some b -> b
      | None ->
        let b = is_const_like info values loc in
        const_like_cache := L.LMap.add loc b !const_like_cache;
        b
    in
    if const_like then
      false
    else if all then
      true
    else
      let not_written_by_closure =
        match L.LMap.find_opt loc !written_by_closure_cache with
        | Some b -> b
        | None ->
          let b = L.LSet.is_empty (written_by_closure info values loc) in
          written_by_closure_cache := L.LMap.add loc b !written_by_closure_cache;
          b
      in
      not not_written_by_closure
end

module With_ALoc =
  Make (Loc_sig.ALocS) (Scope_api.With_ALoc) (Ssa_api.With_ALoc) (Provider_api.ALocProviders)
module With_Loc =
  Make (Loc_sig.LocS) (Scope_api.With_Loc) (Ssa_api.With_Loc) (Provider_api.LocProviders)
include With_ALoc

File: rust_port/crates/flow_env_builder/src/invalidation_api.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::BTreeSet;
use std::hash::Hash;

use dupe::Dupe;
use flow_analysis::scope_api::Def;
use flow_analysis::scope_api::Scope;
use flow_analysis::scope_api::ScopeInfo;
use flow_analysis::ssa_api::Values;
use flow_analysis::ssa_api::WriteLoc;
use flow_data_structure_wrapper::ord_set::FlowOrdSet;

use crate::provider_api::Info as ProviderInfo;
use crate::provider_api::is_provider_state_fully_initialized;

// We consider a binding to be const-like if all reads point to the same
// write, modulo initialization.
pub fn is_const_like<L>(info: &ScopeInfo<L>, values: &Values<L>, loc: &L) -> bool
where
    L: Dupe + Clone + Eq + Ord + Hash,
{
    let Some(def) = info.def_of_use_opt(loc) else {
        return false;
    };
    let def_locs = &def.locs;
    let uses = info.uses_of_use(loc, false);

    // We consider a binding to be const-like if all reads point to the same
    // write, modulo initialization.
    let mut writes: BTreeSet<L> = BTreeSet::new();
    for use_loc in &uses {
        match values.get(use_loc) {
            None => {
                // use is a write
            }
            Some(write_locs) => {
                // use is a read
                // collect writes pointed to by the read, modulo initialization
                for write_loc in write_locs {
                    match write_loc {
                        WriteLoc::Uninitialized => {}
                        WriteLoc::Write(reason) => {
                            writes.insert(reason.loc().dupe());
                        }
                    }
                }
            }
        }
    }

    match writes.iter().next() {
        None => true,
        Some(write_loc) => writes.len() <= 1 && def_locs.iter().any(|def_loc| def_loc == write_loc),
    }
}

#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub enum InitializationValid<L: Dupe + Clone + Eq + Ord> {
    #[default]
    Valid,
    NotWritten {
        possible_generic_escape_locs: FlowOrdSet<L>,
    },
    NullWritten {
        null_provider_loc: L,
        possible_generic_escape_locs: FlowOrdSet<L>,
    },
}

pub fn declaration_validity<L>(
    info: &ScopeInfo<L>,
    values: &Values<L>,
    providers: &ProviderInfo<L>,
    loc: &L,
) -> InitializationValid<L>
where
    L: Dupe + Clone + Eq + Ord + Hash + flow_parser::loc_sig::LocSig,
{
    let Some(def_providers) = providers.providers_of_def(loc) else {
        return InitializationValid::Valid;
    };
    if !def_providers.providers.is_empty()
        && is_provider_state_fully_initialized(&def_providers.state)
    {
        return InitializationValid::Valid;
    }

    let uses = info.uses_of_use(loc, false);

    // Since this variable is not fully initialized, if there are any providers then
    // they must be null providers
    let null_providers: BTreeSet<L> = def_providers
        .providers
        .iter()
        .map(|p| p.reason.loc().dupe())
        .collect();

    let reads_exist = uses.iter().any(|use_loc| values.contains_key(use_loc));

    let possible_generic_escape_locs = def_providers.possible_generic_escape_locs.clone();

    match (reads_exist, !null_providers.is_empty()) {
        (false, _) => InitializationValid::Valid,
        (true, false) => InitializationValid::NotWritten {
            possible_generic_escape_locs,
        },
        (true, true) => {
            let null_provider_loc = null_providers
                .iter()
                .next()
                .expect("null_providers is not empty")
                .dupe();
            InitializationValid::NullWritten {
                null_provider_loc,
                possible_generic_escape_locs,
            }
        }
    }
}

pub fn is_not_captured_by_closure<L>(info: &ScopeInfo<L>, use_loc: &L) -> bool
where
    L: Dupe + Clone + Eq + Ord + Hash,
{
    fn lookup<L>(
        in_current_var_scope: bool,
        info: &ScopeInfo<L>,
        scope: &Scope<L>,
        def: &Def<L>,
    ) -> bool
    where
        L: Dupe + Clone + Eq + Ord + Hash,
    {
        if scope.defs.values().any(|def_prime| def == def_prime) {
            return in_current_var_scope;
        }

        match scope.parent {
            None => true,
            Some(scope_id) => {
                let scope_prime = info.scope(scope_id);
                lookup(
                    in_current_var_scope && scope.lexical,
                    info,
                    scope_prime,
                    def,
                )
            }
        }
    }

    match info.scope_of_use(use_loc) {
        Some((_scope_id, scope)) => match scope.locals.get(use_loc) {
            Some(def) => lookup(true, info, scope, def),
            None => true,
        },
        None => true,
    }
}

pub fn written_by_closure<L>(info: &ScopeInfo<L>, values: &Values<L>, loc: &L) -> BTreeSet<L>
where
    L: Dupe + Clone + Eq + Ord + Hash,
{
    let uses = info.uses_of_use(loc, false);
    let mut result = BTreeSet::new();

    for use_loc in uses {
        match values.get(&use_loc) {
            None => {
                // use is a write
                if !is_not_captured_by_closure(info, &use_loc) {
                    result.insert(use_loc);
                }
            }
            Some(_write_locs) => {
                // use is a read
                // collect writes pointed to by the read, modulo initialization
            }
        }
    }

    result
}

// Some variables are unhavocable by making a function call but still can be havoced in other
// situations. `via_call` indicates whether the havocing operation is a call or something
// else (resetting an activation scope, yielding, entering a new scope)

pub struct InvalidationCaches<L: Clone + Eq + Ord> {
    const_like_cache: BTreeMap<L, bool>,
    written_by_closure_cache: BTreeMap<L, bool>,
}

impl<L: Clone + Eq + Ord> InvalidationCaches<L> {
    pub fn new() -> Self {
        InvalidationCaches {
            const_like_cache: BTreeMap::new(),
            written_by_closure_cache: BTreeMap::new(),
        }
    }
}

impl<L: Clone + Eq + Ord> Default for InvalidationCaches<L> {
    fn default() -> Self {
        Self::new()
    }
}

pub fn should_invalidate<L>(
    all: bool,
    caches: &mut InvalidationCaches<L>,
    info: &ScopeInfo<L>,
    values: &Values<L>,
    loc: L,
) -> bool
where
    L: Dupe + Clone + Eq + Ord + Hash,
{
    let const_like = caches
        .const_like_cache
        .get(&loc)
        .copied()
        .unwrap_or_else(|| {
            let b = is_const_like(info, values, &loc);
            caches.const_like_cache.insert(loc.dupe(), b);
            b
        });

    if const_like {
        return false;
    }

    if all {
        return true;
    }

    let not_written_by_closure = caches
        .written_by_closure_cache
        .get(&loc)
        .copied()
        .unwrap_or_else(|| {
            let b = written_by_closure(info, values, &loc).is_empty();
            caches.written_by_closure_cache.insert(loc.dupe(), b);
            b
        });

    !not_written_by_closure
}
